### PR TITLE
dt_node_info: add the reg size to manage core virtual address map

### PR DESCRIPTION
**Pull request by ldebieve**
A series of changes that use the reg_size from the device tree to manage the call to phys_to_virt which now adds the len management.
Initial patch is to define the size as size_t to be compliant with the different functions using length as size_t and not ssize_t.
Update the driver using the _fdt_fill_device_info to retrieve the reg size and use it to get virtual address.

### CODE REVIEW

File: core/include/kernel/dt.h [hunk 1]
```diff
@@ -199,10 +199,10 @@ static inline paddr_t _fdt_reg_base_address(const void *fdt __unused,
 	return (paddr_t)-1;
 }
 
-static inline ssize_t _fdt_reg_size(const void *fdt __unused,
-				    int offs __unused)
+static inline size_t _fdt_reg_size(const void *fdt __unused,
+				   int offs __unused)
 {
-	return -1;
+	return (size_t)-1;
```
Comment (etienne-lms) [inline core/include/kernel/dt.h:207]: prefer `return DT_INFO_INVALID_REG_SIZE;`

**Review by etienne-lms — COMMENTED** [review]
In commit logs, append parentheses pairs to function labels so that is easier to read-out they are functions: , `...the io_pa_or_va() function.`, `the phys_to_virt() function`.
Also s/mmu/MMU/.

Aside these comments, content looks good to me.

File: core/include/kernel/dt.h (hunk 1 above)
Comment (ldebieve) [reply core/include/kernel/dt.h:207]: Idea here is to be compliant with the previous function that return (paddr_t)-1 too instead of DT_INFO_INVALID_REG

**Comment by jenswi-linaro** [comment]
Why is "core: kernel: use size_t instead of ssize_t for _fdt_reg_size" needed. I'd like to know which problem is solved or avoided.

**Comment by ldebieve** [comment]
@jenswi-linaro This patch is a change to be coherent regarding the functions called later with the result of the _fdt_reg_size.
The functions are core_mmu_add_mapping(), io_pa_or_va(), phys_to_virt() which require size_t len rather than a ssize_t. Except the invalid size than may be return here, there is no need to use a signed type for this size.

**Comment by jenswi-linaro** [comment]
> @jenswi-linaro This patch is a change to be coherent regarding the functions called later with the result of the _fdt_reg_size. The functions are core_mmu_add_mapping(), io_pa_or_va(), phys_to_virt() which require size_t len rather than a ssize_t. Except the invalid size than may be return here, there is no need to use a signed type for this size.

This sounds a bit fluffy. What do you mean by coherent? Is there a problem with the range of a ssize_t compared to a size_t in this context?

**Comment by ldebieve** [comment]
In fact, there is a range issue as the size back from the device tree is defined as a u32 or u64. You cannot limit the range to ssize_t here whereas the DT doesn't limit to signed u32 ot u64. This is probably the best explanation.

**Comment by jenswi-linaro** [comment]
> In fact, there is a range issue as the size back from the device tree is defined as a u32 or u64. You cannot limit the range to ssize_t here whereas the DT doesn't limit to signed u32 ot u64. This is probably the best explanation.

Fair enough, please update the commit message with something along that.

File: core/include/kernel/dt.h (hunk 1 above)
Comment (etienne-lms) [reply core/include/kernel/dt.h:207]: sorry, my comment does not apply to your P-R.
You should  have another patch to update all these casted `-1` to their related `DT_INFO_INVALID_*`, IMO.

**Comment by ldebieve** [comment]
Updated commit comments based on review.

**Comment by jenswi-linaro** [comment]
For the commits "core: kernel: use size_t instead of ssize_t for _fdt_reg_size()" and "core: kernel: add reg_size in the dt_node_info structure":
`Acked-by: Jens Wiklander <jens.wiklander@linaro.org>`

File: core/include/drivers/stm32_i2c.h [hunk 2]
```diff
@@ -48,6 +49,7 @@
 struct stm32_i2c_init_s {
 	unsigned int dt_status;
 	paddr_t pbase;
+	ssize_t reg_size;
```
Comment (etienne-lms) [inline core/include/drivers/stm32_i2c.h:52]: s/ssize_t/size_t/

File: core/include/drivers/stm32_i2c.h [hunk 3]
```diff
@@ -112,6 +115,7 @@ struct i2c_cfg {
  */
 struct i2c_handle_s {
 	struct io_pa_va base;
+	ssize_t reg_size;
```
Comment (etienne-lms) [inline core/include/drivers/stm32_i2c.h:118]: s/ssize_t/size_t/

**Review by etienne-lms — COMMENTED** [review]
Comments for commit "drivers: stm32_i2c: use reg_size to map virtual addresses".

File: core/include/drivers/stm32_i2c.h (hunk 2 above)
Comment (ldebieve) [reply core/include/drivers/stm32_i2c.h:52]: Correct, I'll fix it

File: core/include/drivers/stm32_i2c.h (hunk 3 above)
Comment (ldebieve) [reply core/include/drivers/stm32_i2c.h:118]: ditto

**Comment by ldebieve** [comment]
@etienne-lms Fix pushed

File: core/drivers/stm32_i2c.c [hunk 4]
```diff
@@ -692,11 +692,14 @@ int stm32_i2c_get_setup_from_fdt(void *fdt, int node,
 	memset(init, 0, sizeof(*init));
 
 	_fdt_fill_device_info(fdt, &info, node);
+	assert(info.reg != DT_INFO_INVALID_REG &&
+	       info.reg_size != DT_INFO_INVALID_REG_SIZE &&
+	       info.clock != DT_INFO_INVALID_CLOCK);
```
Comment (jforissier) [inline core/drivers/stm32_i2c.c:697]: Shouldn't `panic()` be used rather than `assert()`? (the latter is a no-op when `-DNDEBUG`).

Comment (ldebieve) [reply core/drivers/stm32_i2c.c:697]: As previously done, we do not expect something wrong from the DT as it comes from the DT SoC level originally, that's why we use for most of our driver this check under assert only in debug mode.

**Review by etienne-lms — COMMENTED** [review]
`Reviewed-by: Etienne Carriere <etienne.carriere@linaro.org>` for the whole series.

File: core/drivers/stm32_i2c.c (hunk 4 above)
Comment (jforissier) [reply core/drivers/stm32_i2c.c:697]: OK

**Comment by ldebieve** [comment]
Squashed fixup commit.
Applied review tag.

**Pull request merged**